Cookbook/Proposals/FeatureSwitchesRedesign.md

```diff
@@ -0,0 +1,222 @@
+
+4. Root.plist should be generated based on the code that declares feature switches.
+
+## Detailed design
+
+### Module specific feature switches
+
+Each type of feature switch will be implemented with its own type to make its semantics clear. We will need `LocalFeatureSwitch` and `RemoteFeatureSwitch` types. When later we introduce a new product configurator service we define a new type: `ProductFeatureSwitch`. They all will have a common `value` property and each provider will have a `get(valueForKey: String)` method and should store their properties in a dictionary.
+
+```swift
+// In BabylonDependencies
+
+public struct RemoteFeatureSwitch<T> {
+    let provider = Current.abTestingService
+    var value: T {
+        return provider.get(valueForKey: self.key)
+    }
+    // Name to be used to generate entries for Root.plist using SwiftSyntax
+    let name: String
+    // key to query the value from provider and key for UserDefaults
+    let key: String
+}
+
+public struct LocalFeatureSwitch<T> {
+    let provider = Current.userDefaults
+    var value: T { ... }
+    let name: String
+    let key: String
+}
+
+public struct ProductFeatureSwitch<T> {
+    let provider = Current.productConfigurator
+    var value: T { ... }
+    let name: String
+    let key: String
+}
+```
+
+
+Using dedicated switch types we will be also able to explicitly specify what remote feature switches are A/B tested and which not:
+
+```swift
+typealias ABTest = RemoteFeatureSwitch
+```
+
+Remote Feature Switches we fetch from Firebase all together and under the hood they are just abstract key-value pairs. So to access them in a module the module should be initialised with this abstract collection of key-value pairs (or access it via `Current` as in examples) so that it can get from it values it cares about and store them for later use.
+
+These feature switches should be separate from static configuration (implemented ATM via `AppConfiguration`) but both should be defined and stored inside the module. This way we break up the monolythic `AppConfiguration` and allow static configuration be configured separately from dynamic feature switches which get their values at runtime.
+
+With patient details situation is a bit different because this data is associated with a `PatientDTO`. We can refactor it so that `PatientDTO` stores abstract key-value pairs and pass it to the frameworks after patient details are retrieved similarly with Firebase remote config (or similarly access it via `Current` when these patient details will be moved there). But we suggest to address this type of feature switches separately as this work depends on other work related to how the app works with patient details. In future these feature switches can become just a new type of switch like `PatientFeatureSwitch` which will use `Current.patient.details` as their provider.
```

Comment:
```suggestion
With patient details situation is a bit different because this data is associated with a `PatientDTO`. We can refactor it so that `PatientDTO` stores abstract key-value pairs and pass it to the frameworks after patient details are retrieved similarly with Firebase remote config (or similarly access it via `Current` when these patient details will be moved there). But we suggest to address this type of feature switches separately/later as this work depends on other work related to how the app works with patient details. In future these feature switches can become just a new type of switch like `PatientFeatureSwitch` which will use `Current.patient.details` as their provider and will expose `Property<Bool>` rather than just `Bool` in its interface.
```

Reply:
@daniloaliberti this is actually mentioned in the doc already, just added a small detail.